<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
        Import Needed Libraries
    </h1>
    </div>

In [68]:
import joblib
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from plotly.offline import iplot , plot
from plotly.subplots import make_subplots
from sklearn.preprocessing import MinMaxScaler , LabelEncoder
from sklearn.model_selection import train_test_split , cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor , GradientBoostingRegressor , BaggingRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error , r2_score
import warnings
warnings.filterwarnings('ignore')

<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
        Reading Dataset And Get Info
    </h1>
    </div>

### Load Data

In [69]:
df = pd.read_csv('Car details v3.csv')

In [70]:
# Show Sample of Data
df.sample(5)

name  year  selling_price  km_driven    fuel  \
160           Toyota Glanza V CVT  2019         894999       9654  Petrol   
3647       Hyundai Santro Xing XL  2006          70000      80000  Petrol   
3556         Hyundai Creta 1.6 SX  2018        1200000      27000  Petrol   
7788  Maruti Wagon R VXI Optional  2017         400000      70000  Petrol   
4332         Maruti Swift AMT ZXI  2018         600000      69779  Petrol   

     seller_type transmission         owner     mileage   engine  max_power  \
160       Dealer    Automatic   First Owner  19.56 kmpl  1197 CC  81.80 bhp   
3647  Individual       Manual  Second Owner         NaN      NaN        NaN   
3556  Individual       Manual   First Owner   15.8 kmpl  1591 CC  121.3 bhp   
7788  Individual       Manual   First Owner  20.51 kmpl   998 CC  67.04 bhp   
4332      Dealer    Automatic   First Owner   22.0 kmpl  1197 CC  81.80 bhp   

              torque  seats  
160   113Nm@ 4200rpm    5.0  
3647             NaN    NaN  
3556  151Nm@ 4850rpm    5.0  
7788   90Nm@ 3500rpm    5.0  
4332  113Nm@ 4200rpm    5.0

In [71]:
# Show Shape of Data
print(f"Number of Row : {df.shape[0]}\nNumber of Columns : {df.shape[1]}")

Number of Row : 8128
Number of Columns : 13


In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


In [73]:
# Check NaN Value
df.isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           222
seats            221
dtype: int64

In [74]:
# Describe Numiric Data
df.describe()

year  selling_price     km_driven        seats
count  8128.000000   8.128000e+03  8.128000e+03  7907.000000
mean   2013.804011   6.382718e+05  6.981951e+04     5.416719
std       4.044249   8.062534e+05  5.655055e+04     0.959588
min    1983.000000   2.999900e+04  1.000000e+00     2.000000
25%    2011.000000   2.549990e+05  3.500000e+04     5.000000
50%    2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    2017.000000   6.750000e+05  9.800000e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000

In [75]:
# Describe non Numiric 
df.describe(exclude=np.number)

name    fuel seller_type transmission        owner  \
count                     8128    8128        8128         8128         8128   
unique                    2058       4           3            2            5   
top     Maruti Swift Dzire VDI  Diesel  Individual       Manual  First Owner   
freq                       129    4402        6766         7078         5289   

          mileage   engine max_power          torque  
count        7907     7907      7913            7906  
unique        393      121       322             441  
top     18.9 kmpl  1248 CC    74 bhp  190Nm@ 2000rpm  
freq          225     1017       377             530

In [76]:
pd.DataFrame({'Count':df.shape[0],
              'Null':df.isnull().sum(),
              'Null %':df.isnull().mean() * 100,
              'Cardinality':df.nunique()
})

Count  Null    Null %  Cardinality
name            8128     0  0.000000         2058
year            8128     0  0.000000           29
selling_price   8128     0  0.000000          677
km_driven       8128     0  0.000000          921
fuel            8128     0  0.000000            4
seller_type     8128     0  0.000000            3
transmission    8128     0  0.000000            2
owner           8128     0  0.000000            5
mileage         8128   221  2.718996          393
engine          8128   221  2.718996          121
max_power       8128   215  2.645177          322
torque          8128   222  2.731299          441
seats           8128   221  2.718996            9

<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
        Handling Missing Data (Nulls)
    </h1>
    </div>

### Handling Missing Data

The number of missing values ​​is small and can be ignored. (Maximum 225 from 8128 Rows)

In [77]:
# Delete Nulls
df.dropna(inplace=True , ignore_index=True)

In [78]:
# Data After Delete Nulls
df.sample(5)

name  year  selling_price  km_driven    fuel  \
3400                   Maruti Alto LX  2006         120000     100000  Petrol   
1189               Mahindra Bolero DI  2008         190000     120000  Diesel   
2160                 Maruti Dzire VXI  2020         580000      70000  Petrol   
4289   Jaguar XF 2.0 Diesel Portfolio  2017        3200000      45000  Diesel   
3347  Ford Figo Aspire 1.5 TDCi Trend  2015         459999      32000  Diesel   

     seller_type transmission         owner     mileage   engine  max_power  \
3400  Individual       Manual  Second Owner   19.7 kmpl   796 CC   46.3 bhp   
1189  Individual       Manual  Second Owner   13.6 kmpl  2523 CC     63 bhp   
2160  Individual       Manual   First Owner  23.26 kmpl  1197 CC  88.50 bhp   
4289      Dealer    Automatic   First Owner  19.33 kmpl  1999 CC    177 bhp   
3347  Individual       Manual  Second Owner  25.83 kmpl  1498 CC     99 bhp   

                      torque  seats  
3400           62Nm@ 3000rpm    5.0  
1189  180 Nm at 1440-1500rpm    7.0  
2160          113nm@ 4400rpm    5.0  
4289     430Nm@ 1750-2500rpm    5.0  
3347     215Nm@ 1750-3000rpm    5.0

In [79]:
# The New of Shape Data
print(f"Number of Row : {df.shape[0]}\nNumber of Columns : {df.shape[1]}")

Number of Row : 7906
Number of Columns : 13


In [81]:
# Check NaN Value
df.isna().sum()

name             0
year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
torque           0
seats            0
dtype: int64

In [82]:
# Describe Numiric (After Delete Null)
df.describe()

year  selling_price     km_driven        seats
count  7906.000000   7.906000e+03  7.906000e+03  7906.000000
mean   2013.983936   6.498137e+05  6.918866e+04     5.416393
std       3.863695   8.135827e+05  5.679230e+04     0.959208
min    1994.000000   2.999900e+04  1.000000e+00     2.000000
25%    2012.000000   2.700000e+05  3.500000e+04     5.000000
50%    2015.000000   4.500000e+05  6.000000e+04     5.000000
75%    2017.000000   6.900000e+05  9.542500e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000

In [83]:
# Describe non Numiric (After Delete Null)
df.describe(exclude=np.number)

name    fuel seller_type transmission        owner  \
count                     7906    7906        7906         7906         7906   
unique                    1982       4           3            2            5   
top     Maruti Swift Dzire VDI  Diesel  Individual       Manual  First Owner   
freq                       129    4299        6563         6865         5215   

          mileage   engine max_power          torque  
count        7906     7906      7906            7906  
unique        393      121       320             441  
top     18.9 kmpl  1248 CC    74 bhp  190Nm@ 2000rpm  
freq          225     1017       377             530

In [84]:
# The New Cardinality
pd.DataFrame({'Count':df.shape[0],
              'Null':df.isnull().sum(),
              'Null %':df.isnull().mean() * 100,
              'Cardinality':df.nunique()
})

Count  Null  Null %  Cardinality
name            7906     0     0.0         1982
year            7906     0     0.0           27
selling_price   7906     0     0.0          670
km_driven       7906     0     0.0          898
fuel            7906     0     0.0            4
seller_type     7906     0     0.0            3
transmission    7906     0     0.0            2
owner           7906     0     0.0            5
mileage         7906     0     0.0          393
engine          7906     0     0.0          121
max_power       7906     0     0.0          320
torque          7906     0     0.0          441
seats           7906     0     0.0            9

<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
         EDA of Data and Get Insights
    </h1>
    </div>

### Types of Car Brands

We will split it into two columns

In [86]:
df[['Brand', 'Model']] = df['name'].str.extract(r'([^ ]+) (.+)', expand=True)


In [92]:
df.head()

year  selling_price  km_driven    fuel seller_type transmission  \
0  2014         450000     145500  Diesel  Individual       Manual   
1  2014         370000     120000  Diesel  Individual       Manual   
2  2006         158000     140000  Petrol  Individual       Manual   
3  2010         225000     127000  Diesel  Individual       Manual   
4  2007         130000     120000  Petrol  Individual       Manual   

          owner     mileage   engine   max_power                    torque  \
0   First Owner   23.4 kmpl  1248 CC      74 bhp            190Nm@ 2000rpm   
1  Second Owner  21.14 kmpl  1498 CC  103.52 bhp       250Nm@ 1500-2500rpm   
2   Third Owner   17.7 kmpl  1497 CC      78 bhp     12.7@ 2,700(kgm@ rpm)   
3   First Owner   23.0 kmpl  1396 CC      90 bhp  22.4 kgm at 1750-2750rpm   
4   First Owner   16.1 kmpl  1298 CC    88.2 bhp     11.5@ 4,500(kgm@ rpm)   

   seats    Brand                   Model  
0    5.0   Maruti         Swift Dzire VDI  
1    5.0    Skoda  Rapid 1.5 TDI Ambition  
2    5.0    Honda      City 2017-2020 EXi  
3    5.0  Hyundai       i20 Sportz Diesel  
4    5.0   Maruti         Swift VXI BSIII

In [91]:
df.drop(columns= ['name'], inplace=True)

In [95]:
print(df['Brand'].nunique())
print(df['Model'].nunique())

31
1982


Drop model column becouse of high cardinality

In [96]:
df.drop(columns=['Model'], inplace=True)

In [98]:
top_20_brands = df['Brand'].value_counts().nlargest(20).reset_index()
top_20_brands.columns = ['Brand', 'Count']  

fig = px.bar(top_20_brands,
             x='Brand',
             y='Count',
             title='Top 20 Car Brands',
             labels={'Count': 'Number of Cars', 'Brand': 'Car Brand'},
             color='Brand'
             )

fig.update_layout(
)

fig.show()

#### Total Selling Price per Years

In [104]:
sum_Price_Years = df.groupby('year')['selling_price'].sum()
px.line(sum_Price_Years,
        x = sum_Price_Years.index,
        y = sum_Price_Years.values,
        labels={'y':'Total Selling Price','year':'Year of Selling the Car'},
        template = 'plotly_dark',
        markers = True,
        line_shape = 'spline',
        color_discrete_sequence = ['red'],
        log_x = True)

In [105]:
print(f"Best year for total car sales profit in the Market in '{sum_Price_Years.idxmax()}' is ${sum_Price_Years.values.max():,.2f}")

Best year for total car sales profit in the Market in '2019' is $1,035,982,984.00


## Does the type of fuel affect on the Selling Price?

In [106]:
type_of_fuel = df['fuel'].value_counts()
px.pie(title='Types of Fuel',
       values = type_of_fuel.values,
       names = type_of_fuel.index,
       template = 'plotly_dark'
       ).update_traces(textinfo='label+percent')

In [107]:
price_Type_Fuel = df.groupby('fuel')['selling_price'].mean()
px.bar(x = price_Type_Fuel.index,
       y = price_Type_Fuel.values,
       template = 'plotly_dark',
       color = price_Type_Fuel.index,
       text_auto = True,
       title = 'Fuel type affects selling Price',
       labels = {'y':'Total Selling Price','x':'Type of Fuel'}
       )

- Diesel and Petrol fuel are the most popluar in the data
- price goes higher with Diesel fuel

## Which is better transmission ,Type of owner and Type Seller Type for the selling price?

### First -> Type of transmission on Selling Price

In [108]:
type_transmission = df['transmission'].value_counts()
px.pie(title='Types of Transmission',
       values = type_transmission.values,
       names = type_transmission.index,
       template = 'plotly_dark'
       ).update_traces(textinfo='label+percent')

- Manual Cars are most popular tham automatic cars

In [109]:
Selling_price_transmission = df.groupby('transmission')['selling_price'].mean()
px.bar(x = Selling_price_transmission.index,
       y = Selling_price_transmission.values,
       template = 'plotly_dark',
       color = Selling_price_transmission.index,
       text_auto = True,
       title = 'Transmission type affects selling Price',
       labels = {'y':'Total Selling Price','x':'Type of Transmission'}
       )

### Second -> Type of Seller on Selling Price

In [110]:
seller_types = df['seller_type'].value_counts()
px.pie(title = 'Seller type affects selling Price',
       values = seller_types.values,
       names = seller_types.index,
       template = 'plotly_dark'
       ).update_traces(textinfo='label+percent')

- Individual sellers are most popular than other sellers

In [111]:
Selling_price_seller_types = df.groupby('seller_type')['selling_price'].mean()
px.bar(x = Selling_price_seller_types.index,
       y = Selling_price_seller_types.values,
       template = 'plotly_dark',
       color = Selling_price_seller_types.index,
       text_auto = True,
       title = 'Seller type affects selling Price',
       labels = {'y':'Total Selling Price','x':'Type of Seller'}
       )

### Third -> Type of Owner on Selling Price

In [112]:
owner_types = df['owner'].value_counts()
px.pie(title = 'Seller type affects selling Price',
       values = owner_types.values,
       names = owner_types.index,
       template = 'plotly_dark'
       ).update_traces(textinfo='label+percent')

- First Owner are most popular than other Owners

In [113]:
Selling_price_owner_types = df.groupby('owner')['selling_price'].mean()
px.bar(x = Selling_price_owner_types.index,
       y = Selling_price_owner_types.values,
       template = 'plotly_dark',
       color = Selling_price_owner_types.index,
       text_auto = True,
       title = 'Owner type affects selling Price',
       labels = {'y':'Total Selling Price','x':'Type of Owner'}
       )

## Total Insights :
- Diesel and Petrol fuel are the most popluar in the data
- Individual sellers are most popular than other sellers
- First Owner are most popular than other Owners
- Manual Cars are most popular tham automatic cars
- price goes higher with Diesel fuel, automatic transmission, Dealer Seller and test Drive Car

<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
         Data Preprocessing
    </h1>
    </div>

## Handling Object DataType

### First Remove any object after numiric data

In [114]:
df = df.replace({'mileage':'[A-Za-z/]','engine':'[A-Za-z]','max_power':'[A-Za-z]'},'',regex=True)
df['mileage'] = df['mileage'].astype(float)
df['engine'] = df['engine'].astype(float)
df['max_power'] = df['max_power'].astype(float)

### Delete the column 'torque' because there are many words and specifications that can be ignored.

In [115]:
df.drop('torque' , axis = 1 , inplace = True)

### Change column year to Age of production Car

In [116]:
df['year'] = 2025 - df['year']
df.rename(columns={'year':'age'},inplace=True)

In [117]:
# Show Data After some of Preprocessing
df.head()

age  selling_price  km_driven    fuel seller_type transmission  \
0   11         450000     145500  Diesel  Individual       Manual   
1   11         370000     120000  Diesel  Individual       Manual   
2   19         158000     140000  Petrol  Individual       Manual   
3   15         225000     127000  Diesel  Individual       Manual   
4   18         130000     120000  Petrol  Individual       Manual   

          owner  mileage  engine  max_power  seats    Brand  
0   First Owner    23.40  1248.0      74.00    5.0   Maruti  
1  Second Owner    21.14  1498.0     103.52    5.0    Skoda  
2   Third Owner    17.70  1497.0      78.00    5.0    Honda  
3   First Owner    23.00  1396.0      90.00    5.0  Hyundai  
4   First Owner    16.10  1298.0      88.20    5.0   Maruti

In [118]:
# After Change DataType of Features
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7906 entries, 0 to 7905
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            7906 non-null   int64  
 1   selling_price  7906 non-null   int64  
 2   km_driven      7906 non-null   int64  
 3   fuel           7906 non-null   object 
 4   seller_type    7906 non-null   object 
 5   transmission   7906 non-null   object 
 6   owner          7906 non-null   object 
 7   mileage        7906 non-null   float64
 8   engine         7906 non-null   float64
 9   max_power      7906 non-null   float64
 10  seats          7906 non-null   float64
 11  Brand          7906 non-null   object 
dtypes: float64(4), int64(3), object(5)
memory usage: 741.3+ KB


### Encode Object DataType 

In [119]:
column_to_encode = ['Brand','fuel','seller_type','transmission','owner']
le = LabelEncoder()
for column in column_to_encode:
    df[column] = le.fit_transform(df[column])

In [120]:
# Show Data After some of Encode
df.head()

age  selling_price  km_driven  fuel  seller_type  transmission  owner  \
0   11         450000     145500     1            1             1      0   
1   11         370000     120000     1            1             1      2   
2   19         158000     140000     3            1             1      4   
3   15         225000     127000     1            1             1      0   
4   18         130000     120000     3            1             1      0   

   mileage  engine  max_power  seats  Brand  
0    23.40  1248.0      74.00    5.0     20  
1    21.14  1498.0     103.52    5.0     26  
2    17.70  1497.0      78.00    5.0     10  
3    23.00  1396.0      90.00    5.0     11  
4    16.10  1298.0      88.20    5.0     20

In [121]:
# Final Info after Encode
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7906 entries, 0 to 7905
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            7906 non-null   int64  
 1   selling_price  7906 non-null   int64  
 2   km_driven      7906 non-null   int64  
 3   fuel           7906 non-null   int64  
 4   seller_type    7906 non-null   int64  
 5   transmission   7906 non-null   int64  
 6   owner          7906 non-null   int64  
 7   mileage        7906 non-null   float64
 8   engine         7906 non-null   float64
 9   max_power      7906 non-null   float64
 10  seats          7906 non-null   float64
 11  Brand          7906 non-null   int64  
dtypes: float64(4), int64(8)
memory usage: 741.3 KB


## Assign feature and target variables

In [122]:
X = df.drop('selling_price' , axis = 1)
y = df['selling_price']

## Splitting Data 

In [123]:
X_train , X_test , y_train , y_test = train_test_split(X , y , test_size = 0.25 , random_state = 44 , shuffle = True)

In [124]:
print(f'Shape of X_Train {X_train.shape}')
print(f'Shape of X_Test {X_test.shape}')
print(f'Shape of Y_Train {y_train.shape}')
print(f'Shape of Y_Test {y_test.shape}')

Shape of X_Train (5929, 11)
Shape of X_Test (1977, 11)
Shape of Y_Train (5929,)
Shape of Y_Test (1977,)


<div style="color:white;
            display:fill;
            border-radius:15px;
            background-color:black;
            font-size:100%;
            font-family:Verdana;
            letter-spacing:1px">
    <h1 style='padding: 20px;
              color:white;
              text-align:center;'>
         Modeling
    </h1>
    </div>

## Use Different Models Algorithm

In [125]:
def Kfold(model,model_name):
    model = cross_val_score(model , X , y , cv = 10)
    model_score = np.average(model)
    print(f"{model_name} score on cross validation: {model_score * 100}%")

def train(model,model_name):
    model.fit(X_train,y_train)
    model_train_score = model.score(X_train,y_train)
    model_test_score = model.score(X_test,y_test)
    print(f"{model_name} model score on Training data: {model_train_score * 100}%\n{model_name} model score on Testing data: {model_test_score * 100}%")

def r2(model,model_name):
    score = r2_score(y_test , model.predict(X_test))
    print(f"R2 Score for {model_name} is {score * 100}%")

## Linear Regression Model

In [126]:
lr = LinearRegression()
Kfold(lr,'Linear Regression')
train(lr,'Linear Regression')
r2(lr,'Linear Regression')

Linear Regression score on cross validation: 66.00283903822536%
Linear Regression model score on Training data: 67.61174801974936%
Linear Regression model score on Testing data: 68.83581706676095%
R2 Score for Linear Regression is 68.83581706676095%


## Random Forest Model

In [127]:
rf_model = RandomForestRegressor(n_estimators = 100 , max_depth = 10)
Kfold(rf_model, "Random Forest")
train(rf_model, "Random Forest")
r2(rf_model, "Random Forest")

Random Forest score on cross validation: 96.32054653659124%
Random Forest model score on Training data: 98.83779333902567%
Random Forest model score on Testing data: 97.09824867933034%
R2 Score for Random Forest is 97.09824867933034%


## KNN

In [128]:
knn = KNeighborsRegressor()
Kfold(knn,'KNN')
train(knn,'KNN')
r2(knn,'KNN')

KNN score on cross validation: 79.96173785489884%
KNN model score on Training data: 86.79608828404307%
KNN model score on Testing data: 82.55109300514209%
R2 Score for KNN is 82.55109300514209%


## Decision Tree Regressor

In [129]:
DTR = DecisionTreeRegressor()
Kfold(DTR,'Decision Tree')
train(DTR,'Decision Tree')
r2(DTR,'Decision Tree')

Decision Tree score on cross validation: 94.55935545344539%
Decision Tree model score on Training data: 99.9662127253933%
Decision Tree model score on Testing data: 96.08683808593746%
R2 Score for Decision Tree is 96.08683808593746%


## XGBoost

In [130]:
xgboost = XGBRegressor()
Kfold(xgboost, "XGBoost")
train(xgboost, "XGBoost")
r2(xgboost, "XGBoost")

XGBoost score on cross validation: 96.24297320842743%
XGBoost model score on Training data: 99.6150553226471%
XGBoost model score on Testing data: 97.44647741317749%
R2 Score for XGBoost is 97.44647741317749%


- Best Model is XGBoost with accuracy 97.4%

## Save rf_model Model 

In [132]:
joblib.dump(rf_model,'rf_model.sav')

['rf_model.sav']